File: map_python/src/Path/path_gate.py
from dataclasses import dataclass, field
from enum import StrEnum

from .path_id import PathId


class GateState(StrEnum):
    OPEN = "open"
    CLOSED = "closed"
    CONDITIONAL = "conditional"


@dataclass(slots=True)
class GateCondition:
    name: str
    satisfied: bool


@dataclass(slots=True)
class PathGate:
    state: GateState
    incoming_path: PathId
    outgoing_path: PathId
    message: str | None = None
    conditions: list[GateCondition] = field(default_factory=list)

File: map_ts/src/Path/path-gate.ts
import type { PathId } from "./path-id.js";

export type GateState =
  | "open"
  | "closed"
  | "conditional";


export type GateCondition = {
  name: string;
  satisfied: boolean;
};


export type PathGate = {
  state: GateState;

  incomingPath: PathId;
  outgoingPath: PathId;

  message?: string;

  conditions: GateCondition[];
};